apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/DefaultUpload.kt

```diff
@@ -1,70 +1,87 @@
 package com.apollographql.apollo3.api
 
+import com.apollographql.apollo3.annotations.ApolloDeprecatedSince
 import okio.BufferedSink
 import okio.BufferedSource
 import okio.ByteString
-import okio.ByteString.Companion.encodeUtf8
-import okio.ByteString.Companion.toByteString
+import okio.FileSystem
+import okio.Path
+import okio.buffer
 import okio.use
+import okio.utf8Size
 
 /**
- * An [Upload] that writes data from the provided content
- *
- * If the content is a [bufferedSource], the [DefaultUpload] will close it once uploaded
- * If the content is a [byteString], the [DefaultUpload] can be reused
+ * A default [Upload] that can upload from a wide variety of content
  */
 class DefaultUpload internal constructor(
-    private val bufferedSource: BufferedSource?,
-    private val byteString: ByteString?,
+    private val writeTo: (BufferedSink) -> Unit,
     override val contentType: String,
     override val contentLength: Long,
     override val fileName: String?,
 ) : Upload {
-  private var bufferedSourceConsumed = false
-
   override fun writeTo(sink: BufferedSink) {
-    if (bufferedSource != null) {
-      check(!bufferedSourceConsumed) { "Apollo: DefaultUpload body can only be read once. If you want to read it several times for logging or other purposes, either buffer it in memory or use your own `Upload` implementation." }
-      bufferedSource.use {
-        sink.writeAll(it)
-      }
-      bufferedSourceConsumed = true
-    } else if (byteString != null) {
-      sink.write(byteString)
-    } else {
-      error("No upload content found")
+    writeTo.invoke(sink)
+  }
+
+  fun newBuilder(): Builder {
+    val builder = Builder()
+        .content(writeTo)
+        .contentType(contentType)
+        .contentLength(contentLength)
+
+    if (fileName != null) {
+      builder.fileName(fileName)
     }
+    return builder
   }
 
   class Builder {
-    private var bufferedSource: BufferedSource? = null
-    private var byteString: ByteString? = null
+    private var writeTo: ((BufferedSink) -> Unit)? = null
     private var contentType: String? = null
     private var contentLength: Long = -1
     private var fileName: String? = null
-    private val hasContent: Boolean
-      get() = bufferedSource != null || byteString != null
 
+    @Deprecated("This API is dangerous because the resulting upload can only be used once and can also lead to resource leaks.",
+        ReplaceWith("content {sink ->\nval source = openSource()\nsource.use {sink.writeAll(it)}\n}"))
+    @ApolloDeprecatedSince(ApolloDeprecatedSince.Version.v3_3_3)
     fun content(content: BufferedSource): Builder = apply {
```

Comment:
This method has a lot of potential foot guns so I marked it as deprecated but it might be going a bit too far. I'm fine keeping it too.